src/firmware/shared/circular_buffer.c

```diff
@@ -0,0 +1,68 @@
+#include "firmware/shared/circular_buffer.h"
+
+#include <assert.h>
+
+struct CircularBuffer
+{
+    int head;
+    int tail;
+    size_t max_size;
+    float *buffer_data;
+    bool isFull;
+};
+
+CircularBuffer_t *circular_buffer_create(size_t size)
+{
+    CircularBuffer_t *new_cbuffer = malloc(sizeof(CircularBuffer_t));
+
+    new_cbuffer->head        = 0;
+    new_cbuffer->tail        = 0;
+    new_cbuffer->max_size    = size;
+    new_cbuffer->buffer_data = (float *)malloc(sizeof(float) * new_cbuffer->max_size);
+    new_cbuffer->isFull      = false;
+
+    return new_cbuffer;
+}
+
+void circular_buffer_destroy(CircularBuffer_t *cbuffer)
+{
+    free(cbuffer);
+}
+
+void circular_buffer_push(CircularBuffer_t *cbuffer, float data)
+{
+    cbuffer->buffer_data[cbuffer->head] = data;
+    cbuffer->head                       = (++cbuffer->head) % cbuffer->max_size;
```

Comment:
nit: just to be slightly more explicit about the precedence of operators here, can we write this as `++(cbuffer->head) % ...` or `(++(cbuffer->head)) % ...`? Just to make it super clear the `++` is operating on the `head`, not the `cbuffer` pointer.

Reply:
Done.